pkg/handler/iam_handler.go

```diff
@@ -0,0 +1,242 @@
+	if h.projectsClient == nil {
+		client, err := resourcemanager.NewProjectsClient(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create the projects client: %w", err)
+		}
+		h.projectsClient = client
+	}
+	return h, nil
+}
+
+// Do removes expired or duplicative IAM bindings added by AOD and adds requested IAM bindings to current IAM policy.
+func (h *IAMHandler) Do(ctx context.Context, r *v1alpha1.IAMRequestWrapper) (nps []*v1alpha1.IAMResponse, retErr error) {
+	for _, p := range r.Request.ResourcePolicies {
+		np, err := h.handlePolicy(ctx, p, r.Duration)
+		if err != nil {
+			retErr = errors.Join(
+				retErr,
+				fmt.Errorf("failed to handle policy update for resource %s: %w", p.Resource, err),
+			)
+		}
+		if np != nil {
+			nps = append(nps, np)
+		}
+	}
+	return
+}
+
+func (h *IAMHandler) handlePolicy(ctx context.Context, p *v1alpha1.ResourcePolicy, ttl time.Duration) (*v1alpha1.IAMResponse, error) {
+	var iamC iamClient
+	switch strings.Split(p.Resource, "/")[0] {
+	case "organizations":
+		iamC = h.organizationsClient
+	case "folders":
+		iamC = h.foldersClient
+	case "projects":
+		iamC = h.projectsClient
+	}
+
+	var np *iampb.Policy
+	b := retry.WithMaxRetries(5, retry.NewFibonacci(500*time.Millisecond))
+	if err := retry.Do(ctx, b, func(ctx context.Context) error {
+		// Get current IAM policy.
+		cp, err := iamC.GetIamPolicy(
+			ctx,
+			&iampb.GetIamPolicyRequest{Resource: p.Resource})
+		if err != nil {
+			return fmt.Errorf("failed to get IAM policy: %w", err)
+		}
+
+		// Update the policy with new IAM binding additions.
+		updatePolicy(cp, p.Bindings, ttl)
+
+		// Set the new policy.
+		setIamPolicyRequest := &iampb.SetIamPolicyRequest{
+			Resource: p.Resource,
+			Policy:   cp,
+		}
+		np, err = iamC.SetIamPolicy(ctx, setIamPolicyRequest)
+		// Retry when set IAM policy fail.
+		if err != nil {
+			return retry.RetryableError(fmt.Errorf("failed to set IAM policy: %w, retrying", err))
+		}
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed to handle IAM request: %w", err)
+	}
+
+	return &v1alpha1.IAMResponse{Resource: p.Resource, Policy: np}, nil
+}
+
+// Remove expired bindings and add or update new bindings with expiration condition.
+func updatePolicy(p *iampb.Policy, bs []*v1alpha1.Binding, ttl time.Duration) {
+	// Clean up current policy.
+	removed := make([]*iampb.Binding, 0, len(p.Bindings))
+	for _, cb := range p.Bindings {
+		// Only clean up AOD bindings.
+		// TODO (#6): Remove expired bindings.
+		if cb.Condition != nil && cb.Condition.Title == conditionTitle {
+			// Exclude duplicative Members from current bindings.
+			ms := make([]string, 0, len(cb.Members))
+			for _, nb := range bs {
+				if cb.Role == nb.Role {
+					for _, m := range cb.Members {
+						if !contains(nb.Members, m) {
+							ms = append(ms, m)
+						}
+					}
+				}
+			}
+			// Copy the bindings with de-dupped members.
+			if len(ms) != 0 {
+				cb.Members = ms
+				removed = append(removed, cb)
+			}
```

Comment:
Maybe use a separate func e.g. `cb.Members = newMembers(cb, bs)`

I'm not a fan of 4 layers of nested for loops. I wonder if we can preprocess `nb` so that it becomes a `map[role]map[member]struct{}` so that you can avoid the nesting the inner two for loops

Reply:
`separate func e.g. cb.Members = newMembers(cb, bs)` -> sg
an issue with preprocess nb to a map is that it may contain bindings that have the same role, I like iterating over the list so that we don't assume anything here.